rflx/expression.py

```diff
@@ -271,10 +280,13 @@ def ada_expr(self) -> ada.Expr:
         return ada.Not(self.expr.ada_expr())
 
     def z3expr(self) -> z3.BoolRef:
-        z3expr = self.expr.z3expr()
-        if isinstance(z3expr, z3.BoolRef):
-            return z3.Not(z3expr)
-        raise TypeError
+        if self._z3boolexpr is None:
+            z3expr = self.expr.z3expr()
+            if not isinstance(z3expr, z3.BoolRef):
+                raise TypeError
```

Comment:
Inconsistent error handling: In some methods a `TypeError` is raised, in others an assertion is used.

Reply:
Fixed in ccf5932.